homework8/task2_/project2_/task2_table_data_sql.py

```diff
@@ -0,0 +1,65 @@
+"""
+Task.
+
+Write a wrapper class TableData for database table, that when initialized with database name and table acts as
+collection object (implements Collection protocol). Assume all data has unique values in 'name' column.
+So, if presidents = TableData(database_name='example.sqlite', table_name='presidents')
+
+then
+--len(presidents) will give current amount of rows in presidents table in database
+--presidents['Yeltsin'] should return single data row for president with name Yeltsin
+--'Yeltsin' in presidents should return if president with same name exists in table
+object implements iteration protocol. i.e. you could use it in for loops::
+--for president in presidents:
+----print(president['name'])
+all above mentioned calls should reflect most recent data. If data in table changed after you created collection
+instance, your calls should return updated data.
+Avoid reading entire table into memory. When iterating through records, start reading the first record,
+then go to the next one, until records are exhausted. When writing tests, it's not always neccessary to mock database
+calls completely. Use supplied example.sqlite file as database fixture file.
+"""
+
+import sqlite3
+from typing import Callable, Iterator, Mapping
+
+
+class TableData(Mapping):
+    def __init__(self, database_name: str, table_name: str) -> None:
+        self.database_name = database_name
+        self.table_name = table_name
+
+    def __enter__(self):
+        self.connection = sqlite3.connect(self.database_name)
+        self.cursor = self.connection.cursor()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.connection.cursor().close()
+        self.connection.close()
+
+    def execute_query(self, query: str, query_args: tuple, fetch: Callable):
+        return fetch(self.cursor.execute(query.format(*query_args)))
+
+    def __len__(self) -> int:
+        return self.execute_query(
+            "SELECT count(*) FROM '{}'",
+            (self.table_name,),
+            lambda cursor: cursor.fetchone()[0],
+        )
+
+    def __getitem__(self, item: str) -> tuple:
+        execution = self.execute_query(
+            "SELECT * FROM '{}' WHERE name = '{}'",
+            (self.table_name, item),
+            lambda cursor: cursor.fetchone(),
+        )
```

Comment:
By inline I mean something like that, you don't need self.execute_query here:
```suggestion
        execution = self.cursor.execute(f"SELECT * FROM '{self.table_name}' WHERE name = '{item}'").fetchone()
```

By the way whats wrong with that code?

Reply:
Maybe some problem here is when cursor is not initialied (no context manager) AttributeError raises but not some, for example, ConnectionError instead? Should we check in all methods as `__len__` ,  `__getitem__` , `__iter__`  firstly check `if self.cursor: execution = ...` and in case not self.cursor raise some ConnectionError?

Reply:
AttributeError is not the best solution but it is ok for that task.

There is another thing in this particular line.

Reply:
I faced with SQL injection here. I tried to managed with it, but it occur again.

Reply:
In the case of the table name, it is ok, but in the case of `{item}` you can deal with it as described in the documentqtion.

Reply:
In documentation:  `cur.execute("select ?", (AUSTRIA, ))`
I did in the same way `f"SELECT * FROM {self.table_name} WHERE name = ?", (item, )` .
It works, but again S608 Possible SQL injection vector through string-based query construction

Reply:
>  It works, but again S608 Possible SQL injection vector through string-based query construction

You still use formating to generate the SQL request. SQLite does not have a proper escaping for table names.
So we need to live with it, nothing is perfect.   So we need to do our best and escape what we can.